c++/src/capnp/any.c++

```diff
@@ -77,4 +79,127 @@ kj::Own<ClientHook> AnyPointer::Pipeline::asCap() {
 
 #endif  // !CAPNP_LITE
 
+StructEqualityResult equal(AnyStruct::Reader left, AnyStruct::Reader right) {
+  auto dataL = left.getDataSection();
+  size_t dataSizeL = dataL.size();
+  while(dataSizeL > 0 && dataL[dataSizeL - 1] == 0) {
+    -- dataSizeL;
+  }
+
+  auto dataR = right.getDataSection();
+  size_t dataSizeR = dataR.size();
+  while(dataSizeR > 0 && dataR[dataSizeR - 1] == 0) {
+    -- dataSizeR;
+  }
+
+  if(dataSizeL != dataSizeR) {
+    return StructEqualityResult::NOT_EQUAL;
+  }
+
+  if(0 != memcmp(dataL.begin(), dataR.begin(), dataSizeL * sizeof(word))) {
+    return StructEqualityResult::NOT_EQUAL;
+  }
+
+  auto ptrsL = left.getPointerSection();
+  auto ptrsR = right.getPointerSection();
+
+  size_t i = 0;
+
+  for(; i < kj::min(ptrsL.size(), ptrsR.size()); i++) {
+    auto l = ptrsL[i];
+    auto r = ptrsR[i];
+    switch(equal(l, r)) {
+    case StructEqualityResult::EQUAL:
+      break;
+    case StructEqualityResult::NOT_EQUAL:
+      return StructEqualityResult::NOT_EQUAL;
+    case StructEqualityResult::UNKNOWN_CONTAINS_CAPS:
+      return StructEqualityResult::UNKNOWN_CONTAINS_CAPS;
+    }
+  }
+
+  return StructEqualityResult::EQUAL;
+
+}
+
+StructEqualityResult equal(AnyList::Reader left, AnyList::Reader right) {
+  if(left.size() != right.size()) {
+    return StructEqualityResult::NOT_EQUAL;
+  }
+  switch(left.getElementSize()) {
+  case ElementSize::VOID:
+  case ElementSize::BIT:
+  case ElementSize::BYTE:
+  case ElementSize::TWO_BYTES:
+  case ElementSize::FOUR_BYTES:
+  case ElementSize::EIGHT_BYTES:
+    if(left.getElementSize() == right.getElementSize()) {
+      if(memcmp(left.getData().begin(), right.getData().begin(), left.getData().size()) == 0) {
+        return StructEqualityResult::EQUAL;
+      } else {
+        return StructEqualityResult::NOT_EQUAL;
+      }
+    } else {
+      return StructEqualityResult::NOT_EQUAL;
+    }
+  case ElementSize::POINTER:
+  case ElementSize::INLINE_COMPOSITE: {
+    auto llist = left.as<List<AnyStruct>>();
+    auto rlist = right.as<List<AnyStruct>>();
+    for(size_t i = 0; i < left.size(); i++) {
+      switch(equal(llist[i], rlist[i])) {
+      case StructEqualityResult::EQUAL:
+        break;
+      case StructEqualityResult::NOT_EQUAL:
+        return StructEqualityResult::NOT_EQUAL;
+      case StructEqualityResult::UNKNOWN_CONTAINS_CAPS:
+        return StructEqualityResult::UNKNOWN_CONTAINS_CAPS;
+      }
+    }
+    return StructEqualityResult::EQUAL;
+  }
+  }
+}
+
+StructEqualityResult equal(AnyPointer::Reader left, AnyPointer::Reader right) {
+  if(right.isCapability()) {
+    return StructEqualityResult::UNKNOWN_CONTAINS_CAPS;
```

Comment:
In fact, _arguably_ we should only use `UNKNOWN_CONTAINS_CAPS` in the case where everything else is equal except for the caps. Otherwise `NOT_EQUAL` is obviously correct. This way a caller could potentially decide that they only care about comparing the non-cap parts, and so could treat `UNKNOWN_CONTAINS_CAPS` as `EQUAL`.